### ISSUE TITLE
Websocket timeout is really two independent timeouts

### ISSUE THREAD

**Issue by easye** [joaotavora/hunchensocket#24]
The `websocket-timeout` parameter in the websocket acceptor is passed to the underlying `hunchentoot` as two independent timeouts for reads and writes on the socket.  If *either* of the timeouts is exceeded, then the socket is closed.  

This confused me in the implementation of a JSON RPC push mechanism <https://github.com/emotiq/emotiq/blob/157484804-create-websocket-interface-20180604a/src/network/websocket/serve.lisp> that would chug along just fine writing messages until it "suddenly stopped" because the read timeout was exceeded.  

To workaround this problem, I set the timeout to an impossibly long 3600 ("an hour").

For a Websocket which is a bi-directional abstraction by default,  the socket timeout should be interpreted as "neither read nor write write activity has occurred" in the given period.

I can't think of an easy way to implement this for `hunchentoot` which is setting the timeout parameter in an implementation-dependent manner presumably thunking down to calls to `setsockopt()` that will signal an error when either timeout is exceeded .  After `hunchentoot` sets the timeout, it does nothing further, and offers no API to cancel or adjust the timeout.

But the documentation for the timeout parameter can note this problem, and this issue will exist for people to stumble across in their hour of need.

**Comment by joaotavora** [comment]
Not sure I understand 100% but I think %99.  I agree there should be two independent timeouts, but if this is impossible to implement, then we have to bite the bullet and manage timers in hunchensocket. Start a timer (cancelling a previous one) when a message is read. Do the same for another write timer. And, like you say, set the socket timeout to something outrageous.  Wonder what thread the errors are signalled in... 
I suggest you try it and PR here :-), there should be something in bordeaux-threads or something to help us out, right?

**Comment by easye** [comment]
The thread errors were due to my slightly unorthodox use of the clients collection on the `websocket-resource`.  In order to implement various networks of publish/subscribe clients, I was calling the write routines on `websocket-client` reference in various collections for pubsub topics.  When the `hunchensocket` removes the `write-lock` mutex, it doesn't set the state of the client to `:disconnected`.  

<https://github.com/joaotavora/hunchensocket/blob/master/hunchensocket.lisp#L179>  

That part can definitely patched to set the state to `:disconnected` if you want to support my sort of use of the client references? 

I just addded a check that `write-lock` isn't `nil` in my unorthodox code.

<https://github.com/emotiq/emotiq/commit/b5a8beb812044f3639660501e1047afb6f5def4e#diff-4298f29498e8a66f9aae811d32e80e26R42>







When `hunchensocket` handles the error from `hunchentoot` on a socket

**Comment by joaotavora** [comment]
So let's just just get some things straight, because I think you're missing different issues here (open a different issue for each... err.. different issue).

This issue is about the failure of hunchensocket to set separate read and write timeouts, right?  You're working around this currently by setting a very large timeout. Correct?

Does the write lock thing that you describe about have anything to do with this issue?

**Comment by easye** [comment]
The write lock problem is a "slight weakness" in `hunchensocket` that might help make it more robust.  Without explictly moving the state of the client away from `:disconnected`, code that has obtained references to a client object may errononeously think the object is in a `:connected` state, but attempts to do I/O will cause the invoking thread to crash.

**Comment by easye** [comment]
>websocket timeout is really two independent timeouts to Not possible to set separate read and write timeouts on the websocket 

The problem is really that the two timeouts are independent:  if either one is true, the socket will be disconnected.

**Comment by joaotavora** [comment]
> The problem is really that the two timeouts are independent: if either one is true, the socket will be disconnected.

You mean they are *interdependent* right? IOW there is only one timeout, call it the "socket activity timeout", and this is insufficient for you, right?

If so, why is it insufficient?

**Comment by easye** [comment]
No,  there isn't only one timeout:  there are separate timers.  If either one goes off, the socket is destroyed.  It *seems* like there is one timeout to the `hunchensocket` user.

```
           SO_SNDTIMEO     set timeout value for output
           SO_RCVTIMEO     set timeout value for input
```

**Comment by joaotavora** [comment]
Ah. But why "it seems"? There is only one `websocket-timeout` field in `websocket-acceptor`:
```
(defclass websocket-acceptor (acceptor)
  ((websocket-timeout :initarg :websocket-timeout
                      :accessor websocket-timeout
                      :initform 300
                      :documentation "Custom WebSocket timeout override."))
  (:default-initargs :request-class 'websocket-request
                     :reply-class 'websocket-reply))
```

Perhaps there could be two, presuming, of course, we could honour their separate semantics, which is what you seem to be telling me in that (slightly cryptic) quote.

**Comment by easye** [comment]
As a user of `hunchensocket` I simply want one timeout.  But `hunchensocket` apes the behavior of `setsockop()` which has two independent timers.

**Comment by joaotavora** [comment]
Ahhhhh. I think I understand.

**Comment by joaotavora** [comment]
So why isn't this a question of rearming the two timeouts everytime a message is read or sent?

**Comment by joaotavora** [comment]
> So why isn't this a question of rearming the two timeouts everytime a message is read or sent? 

(maybe this isn't easy)

**Comment by easye** [comment]
One can't instruct `hunchentoot` to re-arm the timeouts.  That is implementation dependent, if possible at all.

**Comment by joaotavora** [comment]
It should be possible I think, but we would almost surely have to go around hunchentoot directly to the socket. Or we could forget the socket timeout completely and roll our own using timers (which, curiously, is what I suggested in the beginning).

But now that I finally understand the problem, which one of these solutions (or what other solution), do you propose?

**Comment by easye** [comment]
`set-timeouts.lisp` is the file compilation unit in `hunchentoot` involving its timeout API.

**Comment by joaotavora** [comment]
I just remembered that the websocket protocol has a keepalive (ping/pong) mechanism.  Can this be the solution to your problem?

**Comment by easye** [comment]
Sure, one could set a client side timer to push request/response messages across the wire, but that requires that the client explicitly code such logic which isn't as "clean" as one might want.  And such a requirement would marginally impact the performance in the case that one is really trying to pump maximal unidirectional data through the websocket (weak argument against).

Setting the both socket timeouts to an hour seems to be "cleaner" for what I am doing, since in my current use case, there should only ever be one client per Websocket server, so the usual need to cull dangling sockets in order to make a server more efficient isn't present.

**Comment by joaotavora** [comment]
> Sure, one could set a client side timer to push request/response messages across the wire, but that requires that the client explicitly code such logic which isn't as "clean" as one might want. 

No, not explicit. The websocket protocol as a provision for this. From https://tools.ietf.org/html/rfc6455, page 23.  


```5.5.2.  Ping

   The Ping frame contains an opcode of 0x9.

   A Ping frame MAY include "Application data".

   Upon receipt of a Ping frame, an endpoint MUST send a Pong frame in
   response, unless it already received a Close frame.  It SHOULD
   respond with Pong frame as soon as is practical.  Pong frames are
   discussed in Section 5.5.3.

   An endpoint MAY send a Ping frame any time after the connection is
   established and before the connection is closed.

   NOTE: A Ping frame may serve either as a keepalive or as a means to
   verify that the remote endpoint is still responsive.
```

Hunchensocket already replies to ping with pong, so if the other endpoint can send those, you're all set.  I can also code a timer to periodically send ping from hunchensocket to the endpoint (that bit it doesn't do yet, though it will silently ignore a pong if it received one).
